packages/devtools_app/lib/src/screens/memory/panes/diff/controller/item_controller.dart

```diff
@@ -61,29 +75,115 @@ class SnapshotListItem extends DiffListItem with AutoDisposeControllerMixin {
+  final _selectedPathEntry = ValueNotifier<StatsByPathEntry?>(null);
 
   @override
   bool get hasData => heap != null;
 
-  HeapStatistics get statsToShow {
+  HeapClasses classesToShow() {
     final theHeap = heap!;
     final itemToDiffWith = diffWith.value;
-    if (itemToDiffWith == null) return theHeap.stats;
-    return diffStore.compare(theHeap, itemToDiffWith.heap!).stats;
+    if (itemToDiffWith == null) return theHeap.classes;
+    return _diffStore.compare(theHeap, itemToDiffWith.heap!);
   }
 
-  void updateSelectedRecord() => _selectedClassStats.value =
-      statsToShow.statsByClassName[selectedClassName.value];
+  void _handleSelectionChange() {
+    final className = _selectedClass.value;
+    if (className == null) return;
+
+    final heapClasses = classesToShow();
+
+    if (heapClasses is SingleHeapClasses) {
+      _selectedSingleClassStats.value =
+          _selectedClassStats.value = heapClasses.classesByName[className];
+      _selectedDiffClassStats.value = null;
+    } else if (heapClasses is DiffHeapClasses) {
+      _selectedDiffClassStats.value =
+          _selectedClassStats.value = heapClasses.classesByName[className];
+      _selectedSingleClassStats.value = null;
+    } else {
+      throw StateError('Unexpected type: ${heapClasses.runtimeType}.');
+    }
+
+    StatsByPathEntry? newByPathEntry;
+    final path = _selectedPath.value;
+    final classStats = _selectedClassStats.value;
+    if (path != null && classStats != null) {
+      final pathStats = classStats.statsByPath[path];
+      if (pathStats != null) {
+        newByPathEntry = classStats.statsByPathEntries
+            .firstWhereOrNull((e) => e.key == path);
+      }
+    }
+    _selectedPathEntry.value = newByPathEntry;
+  }
+
+  void downloadToCsv() {
+    final csvBuffer = StringBuffer();
+
+    // Write the headers first.
+    csvBuffer.writeln(
+      [
+        'Class',
+        'Library',
+        'Instances',
+        'Shallow Dart Size',
+        'Retained Dart Size',
+        'Short Retaining Path',
+        'Full Retaining Path',
+      ].map((e) => '"$e"').join(','),
+    );
+
+    // // Write a row per retaining path.
+    // final data = heapClassesToShow();
+    // for (var classStats in data.classAnalysis) {
+    //   for (var pathStats in classStats.objectsByPath.entries) {
+    //     csvBuffer.writeln(
+    //       [
+    //         classStats.heapClass.className,
+    //         classStats.heapClass.library,
+    //         pathStats.value.instanceCount,
+    //         pathStats.value.shallowSize,
+    //         pathStats.value.retainedSize,
+    //         pathStats.key.asShortString(),
+    //         pathStats.key.asLongString().replaceAll('\n', ' | '),
+    //       ].join(','),
+    //     );
+    //   }
+    // }
+
+    final file = ExportController().downloadFile(
+      csvBuffer.toString(),
+      type: ExportFileType.csv,
+    );
+
+    notificationService.push(successfulExportMessage(file));
```

Comment:
now that we can access the notifications service from controller code, can we put this upstream into the ExportController.downloadFile method? (can be done in separate PR since there are several other where we'll want to remove existing calls to `notificationService.push(successfulExportMessage(file));`)

Reply:
Added TODO

Reply:
added TODO